pandas/_libs/lib.pyx

```diff
@@ -571,25 +572,43 @@ def maybe_booleans_to_slice(ndarray[uint8_t, ndim=1] mask):
 
 @cython.wraparound(False)
 @cython.boundscheck(False)
-def array_equivalent_object(left: object[:], right: object[:]) -> bool:
+def array_equivalent_object(ndarray left, ndarray right) -> bool:
     """
-    Perform an element by element comparison on 1-d object arrays
+    Perform an element by element comparison on N-d object arrays
     taking into account nan positions.
     """
+    # left and right both have object dtype, but we cannot annotate that
+    #  without limiting ndim.
     cdef:
-        Py_ssize_t i, n = left.shape[0]
+        Py_ssize_t i, n = left.size
         object x, y
+        cnp.broadcast mi = cnp.PyArray_MultiIterNew2(left, right)
+
+    # Caller is responsible for checking left.shape == right.shape
 
     for i in range(n):
-        x = left[i]
-        y = right[i]
+        # Analogous to: x = left[i]
+        x = <object>(<PyObject**>cnp.PyArray_MultiIter_DATA(mi, 0))[0]
+        y = <object>(<PyObject**>cnp.PyArray_MultiIter_DATA(mi, 1))[0]
 
         # we are either not equal or both nan
         # I think None == None will be true here
         try:
             if PyArray_Check(x) and PyArray_Check(y):
-                if not array_equivalent_object(x, y):
+                if x.shape != y.shape:
+                    return False
+                if x.dtype != y.dtype:
```

Comment:
Do we have a test that hits this case (where the shapes are equal but dtypes are not)?

Reply:
good catch.  updated with a test and small change to avoid changing this behavior.